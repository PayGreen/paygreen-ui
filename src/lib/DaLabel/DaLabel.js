import React from 'react';
import PropTypes from 'prop-types';
import {
    colorTypeOptions,
    colorTypeDefault,
    buttonSizeOptions,
    buttonSizeDefault,
} from '../../shared/constants';
import { DaLabelBase } from './style';

const DaLabel = props => {
    return (
        <DaLabelBase {...props}>
            {props.children}
        </DaLabelBase>
    );
};

DaLabel.propTypes = {
    colorType: PropTypes.oneOf(Object.values(colorTypeOptions)),
    fieldSize: PropTypes.oneOf(Object.values(buttonSizeOptions)),
};

DaLabel.defaultProps = {
    colorType: colorTypeDefault,
    fieldSize: buttonSizeDefault,
};

export default DaLabel;
