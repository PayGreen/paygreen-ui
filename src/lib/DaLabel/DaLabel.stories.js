import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, boolean, select } from '@storybook/addon-knobs';
import {
    folder,
    buttonSizeOptions,
    buttonSizeDefault,
    labelHtmlTagOptions,
    labelHtmlTagDefault,
} from '../../shared/constants';
import DaLabel from './DaLabel';

storiesOf(folder.form + 'DaLabel', module)
    .addDecorator(withKnobs)
    .add(
        'DaLabel',
        () => (
            <DaLabel
                fieldSize={select(
                    'Field Size',
                    buttonSizeOptions,
                    buttonSizeDefault,
                )}
                htmlTag={select(
                    'HTML tag',
                    labelHtmlTagOptions,
                    labelHtmlTagDefault,
                )}
                isRequired={boolean('Is required', false)}
            >
                Label Description
            </DaLabel>
        ),
        {
            notes:
                'fieldSize adjusts font size and padding to the size of input/select component associated.',
        },
    );
