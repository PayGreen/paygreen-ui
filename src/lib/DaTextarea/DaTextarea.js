import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import {
    formStatusOptions,
    formStatusDefault,
    inputWidthOptions,
    inputWidthDefault,
    buttonSizeOptions,
    buttonSizeDefault,
} from '../../shared/constants';
import { DaTextareaBase } from './style';

const DaTextarea = props => {
    const [stateValue, setValue] = useState(
        props.value !== undefined ? props.value : '',
    );
    const [stateCharactersStatus, setCharactersStatus] = useState(
        formStatusDefault,
    );

    const {
        hasCounter,
        counterText,
        marginTop,
        marginBottom,
        fieldSize,
        blockWidth,
        ...rest
    } = props;

    const handleChange = e => {
        setValue(e.target.value);
    };

    const charactersLength = stateValue.length || 0;
    useEffect(() => {
        if (charactersLength > props.maxLength * 0.9) {
            setCharactersStatus(formStatusOptions.warning);
        } else if (charactersLength >= props.minLength) {
            setCharactersStatus(formStatusOptions.success);
        }
    }, [charactersLength]);

    const charactersCountBlock = (
        <div>
            <span>{charactersLength}</span>&nbsp;/&nbsp;{props.maxLength}{' '}
            {counterText}
        </div>
    );

    return (
        <DaTextareaBase
            theme={props.theme} // not necessary, only needed for tests
            charactersStatus={stateCharactersStatus}
            inputDisabled={props.disabled}
            inputReadOnly={props.readOnly}
            fieldSize={fieldSize}
            blockWidth={blockWidth}
        >
            <textarea
                {...rest}
                value={stateValue}
                onChange={e => {
                    handleChange(e);
                    props.onChange ? props.onChange(e) : null;
                }}
            />

            {hasCounter && !props.disabled && !props.readOnly
                ? charactersCountBlock
                : null}
        </DaTextareaBase>
    );
};

DaTextarea.propTypes = {
    id: PropTypes.string,
    disabled: PropTypes.bool,
    readOnly: PropTypes.bool,
    minLength: PropTypes.number,
    maxLength: PropTypes.number,
    hasCounter: PropTypes.bool,
    counterText: PropTypes.string,
    blockWidth: PropTypes.oneOf(Object.values(inputWidthOptions)),
    fieldSize: PropTypes.oneOf(Object.values(buttonSizeOptions)),
};

DaTextarea.defaultProps = {
    onChange: undefined,
    minLength: 0,
    maxLength: 2000,
    hasCounter: false,
    counterText: 'characters',
    fieldSize: buttonSizeDefault,
    blockWidth: inputWidthDefault,
};

export default DaTextarea;
