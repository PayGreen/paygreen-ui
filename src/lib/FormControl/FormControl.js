import React, { useState } from 'react';
import { FormControlBase } from './style';

const FormControl = props => {
    const [isOpen, setOpen] = useState(false);

    return (
        <FormControlBase {...props}>
            {React.Children.map(props.children, child =>
                React.cloneElement(child, {
                    isOpen: isOpen,
                }),
            )}
        </FormControlBase>
    );
};

export default FormControl;
