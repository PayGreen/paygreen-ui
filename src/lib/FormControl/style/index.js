import styled from 'styled-components';
import { formStatusDefault } from '../../../shared/constants';
import { statusStyle } from './base';
import { MessageBase } from '../../Message/style';
import { DaLabelBase } from '../../DaLabel/style';
import { DaSelectBase } from '../../DaSelect/style';
import { DaInputBase } from '../../DaInput/style';
import { DaHelpBase } from '../../DaHelp/style';
import { CheckboxGroupBase } from '../../CheckboxGroup/style';
import { RadioGroupBase } from '../../RadioGroup/style';
import { DaTextareaBase } from '../../DaTextarea/style';

const dynamicArea = props => (props.hasHelpButton ? 'help' : 'field');

const FormControlBase = styled.div`
    display: grid;
    grid-template-areas: 'label .' 'field ${dynamicArea}' 'message message';
    grid-template-columns: auto 1fr;

    ${DaLabelBase} {
        grid-area: label;
    }

    ${DaSelectBase}, ${DaInputBase}, ${CheckboxGroupBase}, ${RadioGroupBase}, ${DaTextareaBase} {
        grid-area: field;
        margin-bottom : ${props => props.theme.space.xs};
    }

    ${DaHelpBase} {
        grid-area: help;
    }

    ${MessageBase} {
        display: ${props =>
            props.colorStatus !== formStatusDefault ? 'grid' : 'none'};
        grid-area: message;
    }

    ${props => (props.colorStatus !== formStatusDefault ? statusStyle : null)};

`;

export { FormControlBase };
