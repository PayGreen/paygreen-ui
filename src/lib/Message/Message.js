import React from 'react';
import PropTypes from 'prop-types';
import {
    buttonSizeOptions,
    buttonSizeDefault,
    colorTypeOptions,
    colorTypeDefault,
    colorPalletOptions,
    colorThemeOptions,
    colorThemeDefault,
    formStatusOptions,
    formStatusDefault,
    inputWidthOptions,
    inputWidthDefault,
    blockPositionOptions,
    blockPositionDefault,
    spaceOptions,
    spaceDefault,
} from '../../shared/constants';
import { MessageBase } from './style';

const Message = ({ arrowBlock, children, ...rest }) => {
    const arrow =
        arrowBlock === blockPositionOptions.center
            ? blockPositionOptions.topCenter
            : arrowBlock;

    return (
        <MessageBase {...rest} arrowBlock={arrow}>
            {children}
        </MessageBase>
    );
};

Message.propTypes = {
    colorType: PropTypes.oneOf(Object.values(colorTypeOptions)),
    colorPallet: PropTypes.oneOf([
        colorPalletOptions.theme,
        colorPalletOptions.status,
    ]),
    colorTheme: PropTypes.oneOf(Object.values(colorThemeOptions)),
    colorStatus: PropTypes.oneOf(Object.values(formStatusOptions)),
    blockWidth: PropTypes.oneOf(Object.values(inputWidthOptions)),
    fieldSize: PropTypes.oneOf(Object.values(buttonSizeOptions)),
    hasStaticWidth: PropTypes.bool,
    arrowBlock: PropTypes.oneOf(Object.values(blockPositionOptions)),
    marginTop: PropTypes.oneOf(Object.values(spaceOptions)),
    marginBottom: PropTypes.oneOf(Object.values(spaceOptions)),
};

Message.defaultProps = {
    colorType: colorTypeDefault,
    colorPallet: colorPalletOptions.status,
    colorTheme: colorThemeDefault,
    colorStatus: formStatusDefault,
    blockWidth: inputWidthDefault,
    fieldSize: buttonSizeDefault,
    hasStaticWidth: false,
    arrowBlock: blockPositionDefault,
    marginTop: spaceDefault,
    marginBottom: spaceDefault,
};

export default Message;
