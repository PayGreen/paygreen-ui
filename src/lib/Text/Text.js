import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import {
    textHtmlTagOptions,
    textHtmlTagDefault,

    colorTypeOptions,
    colorTypeDefault,
    colorPalletOptions,
    colorThemeOptions,
    colorThemeDefault,
    greyOptions,
    formStatusOptions,
    formStatusDefault,

    radiusOptions,
    fontSizeOptions,
    fontSizeDefault,
    alignOptions,
    alignDefault,
    spaceOptions,
    spaceDefault,
} from '../../shared/constants';
import { TextBase } from './style';

class Text extends PureComponent {
    render() {
        return <TextBase
            as={this.props.textHtmlTag}
            {...this.props}
        >
            {this.props.children}
        </TextBase>;
    }
}

Text.propTypes = {
    textHtmlTag: PropTypes.oneOf(Object.values(textHtmlTagOptions)),

    colorType: PropTypes.oneOf(Object.values(colorTypeOptions)),
    colorPallet: PropTypes.oneOf(Object.values(colorPalletOptions)),
    colorTheme: PropTypes.oneOf(Object.values(colorThemeOptions)),
    colorWab: PropTypes.oneOf(Object.values(greyOptions)),
    colorStatus: PropTypes.oneOf(Object.values(formStatusOptions)),

    hasBackground: PropTypes.bool,
    radiusSize: PropTypes.oneOf(Object.values(radiusOptions)),
    textSize: PropTypes.oneOf(Object.values(fontSizeOptions)),
    textAlign: PropTypes.oneOf(Object.values(alignOptions)),
    underline: PropTypes.bool,

    paddingLateral: PropTypes.oneOf(Object.values(spaceOptions)),
    paddingTop: PropTypes.oneOf(Object.values(spaceOptions)),
    paddingBottom: PropTypes.oneOf(Object.values(spaceOptions)),

    marginLateral: PropTypes.oneOf(Object.values(spaceOptions)),
    marginTop: PropTypes.oneOf(Object.values(spaceOptions)),
    marginBottom: PropTypes.oneOf(Object.values(spaceOptions)),

    marginInternal: PropTypes.oneOf(Object.values(spaceOptions)),
};

Text.defaultProps = {
    textHtmlTag: textHtmlTagDefault,

    colorType: colorTypeDefault,
    colorPallet: colorPalletOptions.wab,
    colorTheme: colorThemeDefault,
    colorWab: greyOptions.grey60,
    colorStatus: formStatusDefault,

    hasBackground: false,
    radiusSize: radiusOptions.none,
    textSize: fontSizeDefault,
    textAlign: alignDefault,
    underline: false,

    paddingLateral: spaceDefault,
    paddingTop: spaceDefault,
    paddingBottom: spaceDefault,

    marginLateral: spaceDefault,
    marginTop: spaceDefault,
    marginBottom: spaceDefault,

    marginInternal: spaceOptions.sm,
};

export default Text;