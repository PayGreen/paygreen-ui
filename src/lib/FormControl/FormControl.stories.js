import React, { useState } from 'react';
import {
    folder,
    blockPositionOptions,
    buttonSizeOptions,
    fontSizeOptions,
    formStatusOptions,
    formStatusDefault,
    inputWidthOptions,
    spaceOptions,
    spaceDefault,
} from '../../shared/constants';
import labels from '../../shared/labels';
import DaHelp from '../DaHelp/DaHelp';
import DaInput from '../DaInput/DaInput';
import DaLabel from '../DaLabel/DaLabel';
import DaSelect from '../DaSelect/DaSelect';
import DaTextarea from '../DaTextarea/DaTextarea';
import Message from '../Message/Message';
import CheckboxGroup from '../CheckboxGroup/CheckboxGroup';
import Text from '../Text/Text';
import { CardsIcon, QuestionBoldIcon, UsersIcon } from '../Icon/Icon';
import FormControl from './FormControl';

const selectOptions = [
    {
        value: '',
        text: 'Choose your option...',
        disabled: true,
    },
    {
        value: 'first',
        text: 'First option',
    },
    {
        value: 'second',
        text: 'Second option',
    },
    {
        value: 'third',
        text: 'Third option',
    },
];

const checkboxOptions = [
    {
        label: 'First Choice',
        value: 'first',
    },
    {
        label: 'Second Choice',
        value: 'second',
    },
    {
        label: 'Third Choice',
        value: 'third',
    },
];

const MessageContent = (
    <Text textSize={fontSizeOptions.xs}>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aenean felis
        erat, porttitor vel.
    </Text>
);
export default {
    title: folder.form + 'FormControl',
    argTypes: {
        colorStatus: {
            name: labels.colorStatus,
            options: Object.values(formStatusOptions),
            control: 'select',
        },
        required: {
            name: labels.required,
            control: 'boolean',
        },
        disabled: {
            name: labels.disabled,
            control: 'boolean',
        },
        icon: {
            name: labels.icon,
            control: 'boolean',
        },
        hasStaticWidth: {
            name: labels.hasStaticWidth,
            control: 'boolean',
        },
        marginTop: {
            name: labels.marginTop,
            options: Object.values(spaceOptions),
            control: 'select',
        },
        marginBottom: {
            name: labels.marginBottom,
            options: Object.values(spaceOptions),
            control: 'select',
        },
    },
};

export const FormControlWithDaSelect = ({
    colorStatus,
    required,
    marginTop,
    marginBottom,
    icon,
}) => (
    <FormControl
        colorStatus={colorStatus}
        required={required}
        marginTop={marginTop}
        marginBottom={marginBottom}
    >
        <DaLabel>Select label</DaLabel>

        <DaSelect
            defaultValue=""
            options={selectOptions}
            blockWidth={inputWidthOptions.sm}
            icon={icon ? <UsersIcon /> : null}
        />
    </FormControl>
);

FormControlWithDaSelect.args = {
    colorStatus: formStatusDefault,
    required: false,
    marginTop: spaceDefault,
    marginBottom: spaceDefault,
    icon: true,
};

export const FormControlWithDaSelectHelpMessage = ({
    colorStatus,
    required,
    icon,
    hasStaticWidth,
    disabled,
}) => {
    const [isOpen, setOpen] = useState(false);

    return (
        <FormControl
            colorStatus={colorStatus}
            required={required}
            hasStaticWidth={hasStaticWidth}
        >
            <DaLabel>Input label</DaLabel>

            <DaInput
                placeholder="Firstname Lastname"
                blockWidth={inputWidthOptions.sm}
                icon={icon ? <CardsIcon /> : null}
                disabled={disabled}
            />

            <DaHelp onClick={() => setOpen(!isOpen)}>
                <QuestionBoldIcon />
            </DaHelp>

            {isOpen ? (
                <Message
                    arrowBlock={blockPositionOptions.topRight}
                    blockWidth={inputWidthOptions.sm}
                >
                    {MessageContent}
                </Message>
            ) : (
                <></> //to replace 'null' value and avoid error from storybook
            )}
        </FormControl>
    );
};

FormControlWithDaSelectHelpMessage.args = {
    colorStatus: formStatusDefault,
    required: false,
    hasStaticWidth: false,
    disabled: false,
    icon: true,
};

export const FormControlWithCheckboxGroupMessage = ({
    colorStatus,
    required,
}) => (
    <FormControl colorStatus={colorStatus} required={required}>
        <CheckboxGroup
            name="checkboxes"
            legend={'Checkboxes Label'}
            value={checkboxOptions[1].value}
            options={checkboxOptions}
        />

        <Message
            arrowBlock={blockPositionOptions.topLeft}
            fieldSize={buttonSizeOptions.sm}
            blockWidth={inputWidthOptions.xs}
        >
            {MessageContent}
        </Message>
    </FormControl>
);

FormControlWithCheckboxGroupMessage.args = {
    colorStatus: formStatusDefault,
    required: false,
};

export const FormControlWithDaTextarea = ({
    colorStatus,
    required,
    hasStaticWidth,
}) => (
    <FormControl
        colorStatus={colorStatus}
        required={required}
        hasStaticWidth={hasStaticWidth}
    >
        <DaLabel>Textarea description</DaLabel>

        <DaTextarea
            placeholder="Your text&hellip;"
            maxLength={800}
            hasCounter={true}
        />
    </FormControl>
);

FormControlWithCheckboxGroupMessage.args = {
    colorStatus: formStatusDefault,
    required: false,
    hasStaticWidth: false,
};
