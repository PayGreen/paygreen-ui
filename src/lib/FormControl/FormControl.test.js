import React from 'react';
import TestRenderer from 'react-test-renderer';
import { ThemeDefault } from '../../theme';
import DaLabel from '../DaLabel/DaLabel';
import DaTextarea from '../DaTextarea/DaTextarea';
import Message from '../Message/Message';
import Text from '../Text/Text';
import FormControl from './FormControl';

it('renders without crashing', () => {
    const formControl = TestRenderer.create(
        <FormControl theme={ThemeDefault}>
            <DaLabel theme={ThemeDefault}>Label Description</DaLabel>

            <DaTextarea
                theme={ThemeDefault}
                placeholder="Your text&hellip;"
                hasCounter={true}
            />

            <Message theme={ThemeDefault}>
                <Text theme={ThemeDefault}>
                    Lorem ipsum dolor sit amet, consectetur non sodales.
                </Text>
            </Message>
        </FormControl>,
    );
    expect(formControl.toJSON()).toMatchSnapshot();
});
