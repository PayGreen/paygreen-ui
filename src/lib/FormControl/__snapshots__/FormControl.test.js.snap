// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div
  className="style__FormControlBase-sc-1doj43m-0 eXrPaM"
>
  <label
    className="style__DaLabelBase-sc-67p5vh-0 ZaevI"
  >
    Label Description
  </label>
  <div
    className="style__DaTextareaBase-cqgkmw-0 bCAikL"
  >
    <textarea
      disabled={false}
      maxLength={2000}
      minLength={0}
      onChange={[Function]}
      placeholder="Your text…"
      readOnly={false}
      required={false}
      style={
        Object {
          "height": "",
        }
      }
      value=""
    />
    <div>
      <span>
        0
      </span>
       / 
      2000
       
      characters
    </div>
  </div>
  <div
    className="style__MessageBase-ts6lfl-0 dflKhm"
  >
    <p
      className="style__TextBase-sc-11ixwvd-0 kSVwMo"
    >
      Lorem ipsum dolor sit amet, consectetur non sodales.
    </p>
  </div>
</div>
`;
